feat: Persist metadata together with event

patriknw opened the pull request:
* we only used it for ReplicatedEventMetadata and there was no public api to add custom metadata
* adding composite metadata to be able to store more than one metadata entry
* public api to persist event with metadata
* public api to retrieve metadata from EventEnvelope

johanandren reviewed (commented):
Looks reasonable

pvlugter commented on akka-persistence-query/src/main/scala/akka/persistence/query/typed/EventEnvelope.scala:
```diff
@@ -218,11 +219,26 @@ final class EventEnvelope[Event](
+  /**
+   * Java API: The metadata of a given type that is associated with the event.
+   */
+  def getEventMetadata[M](metaDataType: Class[M]): Optional[M] =
```
Not important, but consistently use `metadata` as one word, with `metadataType`?

pvlugter reviewed (commented):
Direction looks good.

patriknw committed:
feat: Persist metadata together with event

* we only used it for ReplicatedEventMetadata and there was no
  public api to add custom metadata
* adding composite metadata to be able to store more than one
  metadata entry
* public api to persist event with metadata
* public api to retrieve metadata from EventEnvelope
* public api to retrieve metadata in EventSourcedBehavior,
  e.g. event handler during replay

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/Effect.scala:
```diff
@@ -192,6 +192,11 @@ trait EffectBuilder[+Event, State] extends Effect[Event, State] {
+  /**
+   * Persist additional metadata together with the event(s).
+   */
+  def persistMetadata(metadata: AnyRef): EffectBuilder[Event, State]
```
One thing to note is that when persisting more than one event the same metadata will be used for all events. Good enough?

johanandren replied:
We can extend if we decide it is needed at some point in the future.

patriknw replied:
changed in https://github.com/akka/akka/pull/32628/commits/dfcbebc414cc2a794295e4578605e4e87da40def

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingSnapshot.scala:
```diff
@@ -179,7 +179,8 @@ private[akka] class ReplayingSnapshot[C, E, S](override val setup: BehaviorSetup
+          eventsReplayed = 0,
+          metadata = None))
     }
 
```
I don't store custom metadata with a snapshot. Do you see a need for that? It would then be accessible from OnRecoverComplete even if no events were replayed.

patriknw replied:
I added it anyway since we already had `metadata: Option[Any]` in the same way in snapshots and it felt more complete, e.g. that you would see the same metadata from RecoverComplete if there were replayed events or just snapshot.
https://github.com/akka/akka/pull/32628/commits/0d7650cfafc3ac39d687c6556e1463fc28ec2353

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/EventSourcedBehavior.scala:
```diff
@@ -115,12 +115,31 @@ object EventSourcedBehavior {
+   * The metadata of the given type that was persisted with an event, if any.
+   * Can only be called from inside the event handler handler or `RecoveryCompleted` of an `EventSourcedBehavior`.
+   */
+  def eventMetadata[M: ClassTag](context: ActorContext[_]): Option[M] = {
```
using the same approach as we have for accessing current sequence number

patriknw reviewed (commented):
I have a few more things to complete, but I think I have covered most of it. Ready for review.

patriknw committed:
NativeImageMetadataSpec

patriknw committed:
include metadata in PublishedEventImpl, and serialization

patriknw committed:
clear _currentMetadata

aludwiko commented on akka-persistence/src/main/scala/akka/persistence/Persistent.scala:
```diff
@@ -246,3 +251,26 @@ private[persistence] final case class PersistentImpl(
+    val metadataType = implicitly[ClassTag[M]].runtimeClass
+    metadata.flatMap {
+      case CompositeMetadata(entries) =>
+        entries.collectFirst {
```
What if I would like to keep 2 strings in the metadata? I thought that CompositeMetadata will be a key-value collection.

johanandren replied:
You'd use a concrete type with two string fields in it, not use a generic/stdlib type directly

patriknw replied:
If key-value then different library code and user code would have to agree on unique keys to not overwrite each other's values accidentally. Easier to use key on the concrete class since that is unique for the specific use of the metadata (as long as it is not a general purpose class).

octonato replied:
But there is also an implicit agreement though. Each library or user code will have to bring its own type and usage of `String` of common/primitive types will lead to bugs. 

We could work with a Seq[Any] and provide a `find` method instead. On the user to decide what is a match.

patriknw replied:
The user would anyway not be able to distinguish a String from another String in such find.

patriknw committed:
EventSourcedMetadataSpec

johanandren commented on akka-persistence/src/main/scala/akka/persistence/Persistent.scala:
```diff
@@ -246,3 +251,26 @@ private[persistence] final case class PersistentImpl(
+        entries.collectFirst {
+          case m: M @unchecked if metadataType.isAssignableFrom(m.getClass) => m
+        }
+      case other: M @unchecked if metadataType.isAssignableFrom(other.getClass) => Some(other)
```
Is the `assignableFrom` flexibility important? I would have thought exactly this class to be enough (and then the entries could be a map and this a map lookup instead).

patriknw replied:
true, CompositeMetadata can probably be a Map with class keys

johanandren reviewed (approved):
LGTM, with one design question

patriknw committed:
store metadata with snapshot

patriknw committed:
exact class match, and some compatibility notes

patriknw committed:
Scala3 fatal warning

patriknw committed:
more mima

octonato commented on akka-persistence-query/src/main/scala/akka/persistence/query/EventEnvelope.scala:
```diff
@@ -59,14 +62,49 @@ final class EventEnvelope(
+  @deprecated("Use metadata with metadataType parameter")
+  def eventMetadata: Option[Any] = {
+    // For backwards compatibility this will use the metadata that was added last (ReplicatedEventMetaData)
+    _eventMetadata match {
+      case Some(CompositeMetadata(entries)) => entries.headOption
+      case s @ Some(_)                      => s
+      case None                             => None
```
nitpicking...


This will look better as

```suggestion
    _eventMetadata.collect {
      case CompositeMetadata(entries) => entries.head
      case s                          => s
    }
```

Assuming of course that a CompositeMetadata has always a head.

patriknw replied:
ok, will change, and see if there are any more places like that. Will also add a validation in CompositeMetadata that it must not be empty

patriknw committed:
collect, review feedback

patriknw committed:
Recovery.default can't be used with Java

johanandren commented on akka-persistence/src/main/resources/META-INF/native-image/com.typesafe.akka/akka-persistence/reflect-config.json:
```diff
@@ -100,8 +100,10 @@
 }, {
   "name" : "akka.persistence.SerializedEvent"
+} , {
+  "name" : "akka.persistence.CompositeMetadata"
```
Is it just the traling space before the comma, I think the spec does a character-by-character compare so probably needs to be exactly as expected, not only valid json. Could be the same thing holds for linefeed at the end as well

patriknw committed:
InternalStableApi for some things used from Akka Projection

aludwiko commented on akka-persistence-typed-tests/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedMetadataSpec.scala:
```diff
@@ -0,0 +1,179 @@
+                    probe ! s"${EventSourcedBehavior.currentMetadata[String](ctx)} onCommand"
+                    Effect
+                      .persist("evt")
+                      .persistMetadata("meta")
```
At first glance, this was a bit confusing for me. How can I persist multiple events with different metadatas? I would expect to pass event + meta together: `Effect.persist("evt", "meta")` I know that this might be problematic for the current API. Maybe we need some form of an additional wrapper? `EventWithMeta(event, meta)`

patriknw replied:
I noted that in https://github.com/akka/akka/pull/32628#discussion_r1925332880
Probably good enough? Would be more API to add but also more difficult to handle in the internal representation because we don't want to wrap in the normal case when there is no metadata.

aludwiko replied:
How about: `Effect.persistWithMeta(event, meta)` ?

patriknw replied:
but it would be for the case when persisting more than one event that it would be needed.

for that it would have to be `Seq[EventWithMeta]` api

patriknw replied:
and it's possible to store more than one metadata entry per event as well
```
Effect
  .persist("evt")
  .persistMetadata("meta1")
  .persistMetadata("meta2")
```

so to fully support that it would be something like:
```
class EventWithMeta[E](event: E, meta: Seq[Any])
```

aludwiko replied:
Yes, but having separate methods would minimize the potential conflict with the current API:
```
.persistWithMeta(event, meta) //for a single one
.persistWithMeta(Seq[EventWithMeta]) //for many
```

patriknw committed:
change api to persistWithMetadata to support different meta on different events

aludwiko reviewed (approved):
LGTM

patriknw committed:
Scala3 error

patriknw committed:
regenerate reflect-config.json
